### PR TITLE
crux-mir branch coverage

### PR DESCRIPTION
**Pull request by spernsteiner**
This branch adds tools to `crux-mir` for measuring and reporting branch coverage of symbolic tests.

To use the coverage measurement system, the user invokes `crux-mir` with the `--branch-coverage`, `--path-sat`, and `--output-directory` options, then runs the new `crux-report-coverage` tool on the recorded coverage data.

The system has three parts.

 1. During translation to Crucible, `crux-mir` records additional information about the translation of branches.  This is used mainly to reconstruct multi-way branches (such as those produced by Rust `match` expressions) from chains of two-way Crucible `Br` instructions.

    Branch translation info is recorded in a separate table, emitted in JSON format after translation.  Crucible `Br`s are associated with table entries by appending the index of the entry to the the source location recorded for the `Br`.  (This is a hack, but there's currently no general mechanism for attaching such metadata to Crucible statements.)

 2. During symbolic execution, an extension of the profiling execution feature emits a new `BRANCH` event the first time each `Br` is reached and emits a `BLOCK` event the first time each basic block is entered.  Currently, each branch exit in Rust leads to a block with no other predecessors, so the `BLOCK` and `BRANCH` information is sufficient to determine which branch exits were taken.

    This approach relies on path-sat checking being enabled so that infeasible branch exits are never executed.  If path-sat is not enabled, the tool will wrongly report that all exits of all branches are covered.

 3. After execution, the user runs `crux-report-coverage`, which processes the branch translation data and the profiling data and reports a warning for any branch exit that was not taken during the test.  The user can provide multiple sets of profiling data at once to check the coverage of an entire test suite as a whole.

Example program:
```Rust
extern crate crucible;
use crucible::*;

#[cfg_attr(crux, crux_test)]
fn crux_test() -> i32 {
    let x = u8::symbolic("x");
    crucible_assume!(x < 2);
    let mut y = 1;
    if x == 0 {
        y += 1;
    } else if x == 1 {
        y += 2;
    } else {
        y += 3;
    }
    y
}
```

Example `crux-report-coverage` output:
```
warning: branch condition never has value false
   ┌─ test/coverage/coverage.rs:11:15
   │
11 │     } else if x == 1 {
   │               ^^^^^^
```

The event recording added to Crucible is language-agnostic, but the `crux-report-coverage` tool is currently tied to `crux-mir` in that it relies on the Rust-specific branch translation data that `crux-mir` produces.  However, it would probably not be difficult to extend `crux-report-coverage` to support other Crux frontends.

### CODE REVIEW

**Review by atomb — APPROVED** [review]
This looks good to me!

File: crux-mir/report-coverage/Cargo.lock [hunk 1]
```diff
@@ -0,0 +1,173 @@
+# This file is automatically @generated by Cargo.
```
[the remark this answers is not in the document]
Comment (spernsteiner) [reply crux-mir/report-coverage/Cargo.lock:1]: The [official recommendation](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries) is to commit `Cargo.lock` for binary packages.  I believe it does get ignored by default (you have to pass `--locked` to use it).

**Comment by brianhuffman** [comment]
@spernsteiner: This PR adds a new argument of type `EventFilter` to function `profilingFeature`, which is used by saw-script. How should I fill in this argument to get saw-script working with the latest crucible?

**Comment by brianhuffman** [comment]
Is `profilingEventFilter` the right value to use?

**Comment by spernsteiner** [comment]
@brianhuffman Yes,  passing `profilingEventFilter` to `profilingFeature` should make it behave just the same as it did before this PR

**Pull request merged**